### ISSUE TITLE
Feature Requests - gg run and tags

### ISSUE THREAD

**Issue by kyle-long** [fidian/gg-core#10]
A project I am working on is currently evaluating tools. There are a couple features that would help us with our use case. I want to know if 1. these features sound like they fit with this tool and 2. (if so) discuss implementation details of these features.

1. A generic `gg run <command>`. This would run an arbitrary command on all the repos. For our specific use case we would also need some sort of a `--how-many-to-run-in-parallel` or some such flag because some of these commands currently need to run in serial.
2. Some sort of a tagging system where I could say "Run this command for all repos tagged with X".

Thoughts?

**Comment by fidian** [comment]
Sounds like a good thing. What I see is three features.

1. `gg run <command>` as you have described.
2. `gg --parallel=3 <action>` or similar to control the number of parallel operations from the command line. This won't be easy as a flag, but could very easily be done through an environment variable like `GG_PARALLEL=1 gg <action>`. It's already part of the config (more below).
3. Filtering repositories by tags. Like the parallel solution above, this would be a flag that applies to the main command (`gg --filter-tag=tag1,tag2,tag3`) or an environment variable (`GG_FILTER_TAG=tag1,tag2,tag3 gg <action>`).

The INI file setting you can currently use for controlling parallelism. This example sets it to 10. Note that this is either `~/.ggconfig` or `project/.gg/config`. It is not the normal manifest file. The syntax is my best guess. I think it's right but I don't have my laptop handy to test it.

```ini
parallel.jobs=10
```

And now we have tags. Were you thinking of adding those to the INI file too? Here's an example that shows a project with one tag and another that has two tags. Repeating the key is weird, but that's just kinda how INI files work, though we could support commas as well, which is shown in the third example.

```ini
[project-name]
remote=git@github.com:example/project.git
path=project-name
tag=tag1

[second-project]
remote=git@github.com:example/second-project.git
path=second-place
branch=develop
tag=tag2
tag=tag3

[config]
remote=git@github.com:example/config.git
path=data/config
tag=tag1,tag2
```

Overall, the three ideas sound good. What are your thoughts about using environment variables instead of options to the `gg` command?

**Comment by kyle-long** [comment]
Hi. I misread the section on parallel config. I thought it was restricted to 1 per core. Yes env variables seem just fine and as for the tags, I was planning on those going into the ini file. I had in mind the last version version of tags in your ini file example but I am open to any way you would like to do it. Which version do you prefer? It sounds like you prefer the repeating tag key?

**Comment by fidian** [comment]
I just looked again and the `PARALLEL_JOBS` environment variable already controls this behavior. Looks like item 2 is already resolved. I think I want to rename it to `GG_PARALLEL_JOBS` because we could be introducing other environment variable things (like filters). Might as well do it now before this project explodes with popularity.

As for the tags, I am unsure if I prefer one or the other. Repeating the tag looks like overwriting a value or it could be confusing for others. The repeated value is already an INI "standard" (assuming one can write those two words together without causing readers to laugh). I think I'll go for the repeated values for now.

I'll poke at the code today and see if I can give you the necessary functionality. Maybe even close some bugs.

**Comment by fidian** [comment]
It turns out that the INI parsing that `gg` uses is different than the INI parsing library I have in `bpm`. This one overwrites. Thus the tags need to be specified as a comma separated list.

The plan was to use `bpm` modules and build a version of these core files from those other libraries. They are far better tested and work faster, but that would take a bit of restructuring to the code. I think I'll keep things as-is for now and set a precedent to use a single string. Assuming I do get to using the `bpm` libraries, I'll make sure to support multiple lines and each having commas.

Anyway, I've updated the code to implement everything you wanted. I would like very much to know how well it works out for you and your team, if you decide to pick this project. Closing the issue because I believe there's no more work to do.
